http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/SocketOptionsConfiguration.java

```diff
@@ -0,0 +1,215 @@
+                                                       "connectTimeoutMs");
+        this.keepAliveIntervalSecs = Validate.isNotNegative(builder.keepAliveIntervalSecs,
+                                                            "keepAliveIntervalSecs");
+        this.keepAliveTimeoutSecs = Validate.isNotNegative(builder.keepAliveTimeoutSecs,
+                                                           "keepAliveTimeoutSecs");
+    }
+
+    /**
+     * @return socket domain
+     */
+    public SocketOptions.SocketDomain domain() {
+        return domain;
+    }
+
+    /**
+     * @return socket type
+     */
+    public SocketOptions.SocketType type() {
+        return type;
+    }
+
+    /**
+     * @return number of milliseconds before a connection will be considered timed out
+     */
+    public int connectTimeoutMs() {
+        return connectTimeoutMs;
+    }
+
+    /**
+     * @return number of seconds between TCP keepalive packets being sent to the peer
+     */
+    public int keepAliveIntervalSecs() {
+        return keepAliveIntervalSecs;
+    }
+
+    /**
+     * @return number of seconds to wait for a keepalive response before considering the connection timed out
+     */
+    public int keepAliveTimeoutSecs() {
+        return keepAliveTimeoutSecs;
+    }
+
+    public static Builder builder() {
+        return new DefaultSocketOptionsConfigurationBuilder();
+    }
+
+    /**
+     * A builder for {@link SocketOptionsConfiguration}.
+     *
+     * <p>All implementations of this interface are mutable and not thread safe.</p>
+     */
+    public interface Builder {
+
+        /**
+         * Sets the socket domain
+         * @param domain socket domain
+         * @return Builder
+         */
+        Builder domain(SocketOptions.SocketDomain domain);
+
+        /**
+         * Sets the socket type
+         * @param type socket type
+         * @return Builder
+         */
+        Builder type(SocketOptions.SocketType type);
+
+        /**
+         * Sets the number of milliseconds before a connection will be considered timed out
+         * @param connectTimeoutMs number of milliseconds before a connection will be considered timed out
+         * @return Builder
+         */
+        Builder connectTimeoutMs(int connectTimeoutMs);
+
+        /**
+         * Sets the number of seconds between TCP keepalive packets being sent to the peer
+         * @param keepAliveIntervalSecs number of seconds between TCP keepalive packets being sent to the peer
+         * @return Builder
+         */
+        Builder keepAliveIntervalSecs(int keepAliveIntervalSecs);
+
+        /**
+         * Sets the number of seconds to wait for a keepalive response before considering the connection timed out
+         * @param keepAliveTimeoutSecs number of seconds to wait for a keepalive response before considering the connection timed out
+         * @return Builder
+         */
+        Builder keepAliveTimeoutSecs(int keepAliveTimeoutSecs);
```

Comment:
Do we expect customers to configure `keepAliveIntervalSecs` and `keepAliveTimeoutSecs`? If not, should we just introduce `tcpKeepAlive(Boolean tcpKeepAlive)` and pass default values defined in the SDK to CRT? Because exposing too many knobs may make the SDK more complex to use.

Reply:
Will default to your judgement there. I suspect just being able to enable keepAlive should be enough.

Reply:
no, keep alive is worthless without setting those values. The kernel defaults are never what anyone needs for these use-cases. So, it needs to be here.

Reply:
My concern is that these settings may be too "advanced", so some customers may just set some random values w/o doing any research. I'm fine with adding these two, but I think we should still introduce a boolean tcpKeepAlive and define some "safe" defaults for  tcpInterval and tcpKeepAliveIntervalSecs if they are not configured. WDYT?

Reply:
IMO I think the right way to do this would be to separate KeepAliveConfiguration into a dedicated object.

If it's not set, that means customers want the default, disabled. If they DO set it, then they should be setting both the interval and the timeout, and they're "opting-in" to knowing what they're doing.

Reply:
It's either add these values or don't bother. The kernel defaults will not meet anyone's needs. It's on the scale of hours before the first probe gets sent.  It IS too advanced, but the answer isn't to do it for the user. By the time you're tuning keep alive, both of these values need to be a conscious decision by the user.

Reply:
@nikp I think a separate class is fine. Can we name it `TcpKeepAliveConfiguration` though? because people may confuse it with HTTP keep alive (we've had customers asking for clarification before)

@JonathanHenson I think that's fair. We'd probably need to provide some guidance on how to tune them.

Reply:
I'm gonna submit a new revision with that approach - I think it'll work nicely :)

Reply:
I submitted TcpKeepAliveConfiguration, please take a look @zoewangg